devices/Sim7080/SimController.cs

```diff
@@ -0,0 +1,261 @@
+            {
+                Debug.WriteLine(exception.Message);
+
+                return ConnectionStatus.Error;
+            }
+        }
+
+        /// <summary>
+        /// Unsubscribe to Cloud-2-Device Topic 
+        /// </summary>
+        /// <param name="serialPort">The UART <see cref="SerialPort"/> for communication with the modem</param>
+        /// <param name="topic">The name of the topic</param>
+        /// <returns><see cref="ConnectionStatus"/></returns>
+        public static ConnectionStatus UnsubscribeFromTopic(SerialPort serialPort, string topic)
+        {
+            try
+            {
+                ExecuteCommand(serialPort, $"AT+SMUNSUB=\"{topic}\"");
+
+                return ConnectionStatus.Connected;
+            }
+            catch (Exception exception)
+            {
+                Debug.WriteLine(exception.Message);
+
+                return ConnectionStatus.Error;
+            }
+        }
+
+        /// <summary>
+        /// Disconnect from MQTT Endpoint
+        /// </summary>
+        /// <param name="serialPort">The UART <see cref="SerialPort"/> for communication with the modem</param>
+        /// <returns><see cref="ConnectionStatus"/></returns>
+        public static ConnectionStatus EndpointDisconnect(SerialPort serialPort)
+        {
+            try
+            {
+                ExecuteCommand(serialPort, "AT+SMDISC");
+
+                return ConnectionStatus.Disconnected;
+            }
+            catch (Exception exception)
+            {
+                Debug.WriteLine(exception.Message);
+
+                return ConnectionStatus.Error;
+            }
+        }
+
+        /// <summary>
+        /// Send message to the serial port
+        /// </summary>
+        /// <param name="serialPort">The UART <see cref="SerialPort"/> for communication with the modem</param>
+        /// <param name="message"></param>
+        /// <param name="topic"></param>
+        /// <returns></returns>
+        public static bool SendMessage(SerialPort serialPort, string message, string topic)
+        {
+            try
+            {
+                //Simcom module MQTT subscribe to D2C topic
+                ExecuteCommand(serialPort, $"AT+SMPUB=\"{topic}\",{message.Length},1,1");
+
+                //Simcom module MQTT sends the message
+                ExecuteCommand(serialPort, message);
+
+                return true;
+            }
+            catch (Exception exception)
+            {
+                Debug.WriteLine(exception.Message);
+
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Extract the response from AT Acknowledgement message
+        /// </summary>
+        /// <param name="responseMessage">The acknowledgement message</param>
+        /// <returns></returns>
+        public static string ExtractATResponse(string responseMessage)
+        {
+            return Regex.Match(responseMessage, @"""([^""]*)""").Groups[1].Value;
```

Comment:
I know Regex are great especially as I put quite a lot of love in the nano implementation. But they are not that fast neither. Have you run tests to see if they are performing enough for your implemenation?

Reply:
Had the same feedback from Jose. Initially had a routine, replaces it with regex and didn't notice any performance issues. Decided to use the regex because it simplifies the code.

Reply:
Then fine for regex :-D

Reply:
The actual ACK is for IPAddress : +CNACT: 0,1,"10.94.36.44" +CNACT: 1,0,"0.0.0.0" +CNACT: 2,0,"0.0.0.0" +CNACT: 3,0,"0.0.0.0" 

The actual ACK is for the Provider: +COPS: 0,0,"CHN-CT",9